src/ast/passes/fold_literals.cpp

```diff
@@ -0,0 +1,450 @@
+    return nullptr;
+  }
+
+  return std::visit(
+      [&](const auto &v) -> Expression * {
+        if constexpr (std::is_same_v<std::decay_t<decltype(v)>, uint64_t>) {
+          return ast_.make_node<Integer>(v, Location(op.loc));
+        } else {
+          return ast_.make_node<NegativeInteger>(v, Location(op.loc));
+        }
+      },
+      result.value());
+}
+
+Expression *LiteralFolder::visit(PositionalParameterCount &param)
+{
+  // This is always an unsigned integer value.
+  return ast_.make_node<Integer>(bpftrace_.num_params(), Location(param.loc));
+}
+
+Expression *LiteralFolder::visit(PositionalParameter &param)
+{
+  // By default, we treat parameters as integer literals if we can, and
+  // rely on the user to have an explicit `str` cast.
+  const std::string &val = bpftrace_.get_param(param.n);
+  // If empty, treat as zero. This is the documented behavior.
+  if (val.empty()) {
+    return ast_.make_node<Integer>(static_cast<uint64_t>(0),
+                                   Location(param.loc));
+  }
+  try {
+    if (val[0] == '-') {
+      auto v = util::to_int(val, 0);
+      return ast_.make_node<NegativeInteger>(v, Location(param.loc));
+    } else {
+      auto v = util::to_uint(val, 0);
+      return ast_.make_node<Integer>(v, Location(param.loc));
+    }
+  } catch (const std::exception &e) {
+    // Treat this as the original string.
+    return ast_.make_node<String>(val, Location(param.loc));
+  }
+}
+
+Expression *LiteralFolder::visit(Call &call)
+{
+  // If this is the string function, then we can evaluate the given literal
+  // as a string (e.g. this covers str(0) and str($3)).
+  if (call.func == "str" && !call.vargs.empty()) {
+    std::string s;
+
+    // First, we need to check if this directly wraps a positional parameter.
+    // If yes, then we prevent it from expanding to zero as is the default.
+    if (auto *param = dynamic_cast<PositionalParameter *>(call.vargs.at(0))) {
+      call.vargs[0] = ast_.make_node<String>(bpftrace_.get_param(param->n),
+                                             Location(param->loc));
+    } else if (auto *binop = dynamic_cast<Binop *>(call.vargs.at(0))) {
+      auto *param = dynamic_cast<PositionalParameter *>(binop->left);
+      if (param && binop->op == Operator::PLUS) {
+        binop->left = ast_.make_node<String>(bpftrace_.get_param(param->n),
+                                             Location(param->loc));
+      }
+    }
+
+    // Now we can expand normally.
+    Visitor<LiteralFolder, Expression *>::visit(call);
+
+    // If this is an integer of some kind, then fold it into a string.
+    if (auto *n = dynamic_cast<Integer *>(call.vargs.at(0))) {
+      std::stringstream ss;
+      ss << n->value;
+      s = ss.str();
+    } else if (auto *n = dynamic_cast<NegativeInteger *>(call.vargs.at(0))) {
+      std::stringstream ss;
+      ss << n->value;
+      s = ss.str();
+      return ast_.make_node<String>(ss.str(), Location(call.loc));
```

Comment:
why don't we need to truncate the string if it's negative?

Reply:
Good catch, this is definitely a mistake. Fixed in my local, will fix on update.